src/extension.ts

```diff
@@ -1,15 +1,106 @@
-      return new GatherProvider(context.extensionPath);
-    },
-  };
+const registeredButtons: vscode.Disposable[] = [];
 
-  return api;
+export async function activate() {
+  try {
+    sendTelemetryEvent(Telemetry.GatherIsInstalled);
+    const cellStatusBarItems = new WeakMap<vscode.NotebookCell, vscode.NotebookCellStatusBarItem>();
+    const gatherProviderMap = new Map<string, IGatherProvider>();
+
+    const jupyter = vscode.extensions.getExtension<IJupyterExtensionApi>(
+      Constants.jupyterExtension
+    );
+
+    if (jupyter) {
+      if (!jupyter.isActive) {
+        await jupyter.activate();
+        await jupyter.exports.ready;
+      }
+
+      vscode.commands.registerCommand(Constants.gatherNativeNotebookCommand, async (cell: vscode.NotebookCell) => {
+        const id = cell.notebook.uri.toString();
+        const provider = gatherProviderMap.get(id);
+
+        if (provider) {
+          const item = cellStatusBarItems.get(cell);
+          if (item) {
+            item.show();
+            await provider.gatherCode(cell, false);
+            item.hide();
+          } else {
+            provider.gatherCode(cell, false);
+          }
+        } else {
+          vscode.window.showInformationMessage(localize.Common.runCells());
+        }
+      });
+
+      vscode.notebook.onDidCloseNotebookDocument((notebook) => {
+        if (gatherProviderMap.has(notebook.uri.toString())) {
+          gatherProviderMap.delete(notebook.uri.toString());
+        }
+      });
+
+      vscode.notebook.onDidOpenNotebookDocument((notebook) => {
+        notebook.cells.forEach(cell => {
+          const item = cellStatusBarItems.get(cell) ?? vscode.notebook.createCellStatusBarItem(cell, vscode.NotebookCellStatusBarAlignment.Right);
+          cellStatusBarItems.set(cell, item);
+          item.text = 'Gathering $(sync~spin)';
+          item.hide();
+        });
+      });
+
+      const button = jupyter.exports.registerCellToolbarButton(
+        async (cell: vscode.NotebookCell, isInteractive: boolean, resource: vscode.Uri) => {
+          const provider = gatherProviderMap.get(resource.toString());
+          if (provider) {
+            provider.gatherCode(cell, isInteractive);
+          } else {
+            vscode.window.showInformationMessage(localize.Common.runCells());
+          }
+        },
+        'gather',
+        [vscode.NotebookCellRunState.Success],
+        localize.Common.gatherTooltip());
+      registeredButtons.push(button);
+
+      jupyter.exports.onKernelStateChange((nbEvent: KernelStateEventArgs) => {
+        if (nbEvent.state === KernelState.started) {
+          let language = Constants.PYTHON_LANGUAGE;
+          vscode.notebook.visibleNotebookEditors.forEach((ne) => {
+            if (ne.document.uri.toString() === nbEvent.resource.toString() && ne.document.languages[0]) {
+              language = ne.document.languages[0];
+            }
+          });
+
+          let provider = new GatherProvider(language);
```

Comment:
Do we need to get rid of the provider somewhere? Like on kernel shutdown?

Reply:
On kernel shutdown would have to wait until we release the other ext, it'd need changes on the API there.
I could do it when this extension deactivates. What do you think?

Reply:
We talked offline. Seems okay. A new one gets created on every kernel start. We might want to destroy them on document close, but I don't think we need to.